File: packages/core/stages/sf_ocr/ocr.ts
import { $, spawn } from "bun";
import { existsSync, readFileSync, rmSync } from "node:fs";
import { join, resolve } from "node:path";
import { ensureDir } from "@repo/util/file_op";
import { emitLog, nowISO } from "@repo/core/stages/utils/utils.ts";
import { TaskCtx, setStage } from "@repo/core/context/context.ts";
import { startLog } from "../utils/log.ts";
import { OcrFramesResult } from "@repo/subtitle-ocr/types";
import { REPO_ROOT } from "@repo/config/root";
import { log } from "@repo/util/log";

// 关键帧 OCR：消费 sf_ocr_pre 落盘的关键帧（`{start_ms}_{end_ms}.png`），
// 经 subtitle-ocr CLI --dir 批量识别（ms_ms 双时刻 → 每张产出 start/end 两个结果），
// 只写出 `<taskDir>/sf_ocr/frames.json`（OcrFramesResult 原始逐帧结果）。
// 段合并 / 时间调整 / LLM 修正由下游 sf_ocr_fix stage 负责。
export async function stageSfOcr(ctx: TaskCtx) {
  const taskId = ctx.task.id;
  const taskDir = ctx.task.task_dir;
  startLog(ctx.task.current_stage, ctx.task.id);
  await setStage(taskDir, "sf_ocr", {
    last_message: "OCR'ing keyframes...",
    progress: 0,
  });

  const preDir = resolve(taskDir, "sf_ocr_pre", "frames");
  if (!existsSync(preDir)) {
    throw new Error(`Keyframe dir not found: ${preDir} — run sf_ocr_pre first`);
  }

  const ocrArgs = ctx.input.stages.sf_ocr;

  const ocrBin = join(REPO_ROOT, "target", "release", "subtitle-ocr");
  if (!existsSync(ocrBin)) {
    log(`[sf_ocr] subtitle-ocr 未构建，自动编译...`);
    const build = await $`cargo build --release -p subtitle-ocr-cli --bin subtitle-ocr`
      .cwd(REPO_ROOT)
      .nothrow();
    if (build.exitCode !== 0) {
      throw new Error(`subtitle-ocr 编译失败 (exit ${build.exitCode}):\n${build.stderr}`);
    }
  }

  const sfOcrDir = resolve(taskDir, "sf_ocr");
  const outFile = join(sfOcrDir, "frames.json");

  const args = [
    "--dir",
    preDir,
    "--out",
    outFile,
    "--text-confidence-threshold",
    String(ocrArgs.text_confidence_threshold),
    ...(ocrArgs.subtitleOnly ? ["--subtitle-only"] : []),
  ];
  log(`subtitle-ocr ${args.join(" ")}`);
  const proc = spawn([ocrBin, ...args], {
    cwd: REPO_ROOT,
    stdout: "inherit",
    stderr: "inherit",
  });
  const exitCode = await proc.exited;
  if (exitCode !== 0) {
    throw new Error(`subtitle-ocr failed with exit code ${exitCode}`);
  }

  const data = JSON.parse(readFileSync(outFile, "utf-8")) as OcrFramesResult;
  const frameResults = data.frames ?? [];
  if (!frameResults.length) {
    throw new Error("sf_ocr: no OCR results from keyframes");
  }
  log(`${frameResults.length} frame results -> ${outFile}`);

  if (ocrArgs?.cleanupFrames) {
    rmSync(preDir, { recursive: true, force: true });
    log(`Keyframes cleaned up`);
  }

  await setStage(taskDir, "sf_ocr", {
    status: "success",
    completed_at: nowISO(),
    progress: 100,
    last_message: `OCR'd ${frameResults.length} frame results`,
  });
}

File: packages/core/src/stages/sf_ocr/ocr.rs
//! sf_ocr: 关键帧 OCR, 消费 sf_ocr_pre 落盘的关键帧, 经 subtitle-ocr CLI 批量识别。
//!
//! 镜像 TS `packages/core/stages/sf_ocr/ocr.ts` (stageSfOcr)。
//! 只写出 `<taskDir>/sf_ocr/frames.json` (OcrFramesResult 原始逐帧结果);
//! 段合并 / 时间调整 / LLM 修正由下游 sf_ocr_fix 负责。

use crate::context::TaskCtx;
use crate::stages::sf_ocr::args::SfOcrArgs;
use crate::stages::utils::{
    StagePatch, StageStatus, cargo_build_bin, find_release_bin, now_iso,
    set_stage_anyhow, sf_ocr_dir, sf_ocr_pre_dir,
};
use std::process::Command;

/// 读取 sf_ocr 配置 (缺省用 SfOcrArgs::default)。
fn read_args(ctx: &TaskCtx) -> SfOcrArgs {
    ctx.input
        .get("stages")
        .and_then(|v| v.get("sf_ocr"))
        .and_then(|v| serde_json::from_value(v.clone()).ok())
        .unwrap_or_default()
}

/// 入口 (镜像 TS `stageSfOcr`)。
pub fn stage_sf_ocr(ctx: &TaskCtx) -> anyhow::Result<()> {
    let task_dir = ctx.task.task_dir.clone();
    tracing::info!(target: "sf_ocr", "sf_ocr: start");

    set_stage_anyhow(
        &task_dir,
        "sf_ocr",
        StagePatch {
            last_message: Some("OCR'ing keyframes...".into()),
            progress: Some(0.0),
            ..Default::default()
        },
    )?;

    let cfg = read_args(ctx);

    let frame_dir = sf_ocr_pre_dir(&task_dir).join("frames");
    if !frame_dir.exists() {
        return Err(anyhow::anyhow!(
            "Keyframe dir not found: {} — run sf_ocr_pre first",
            frame_dir.display()
        ));
    }

    let bin = match find_release_bin("subtitle-ocr") {
        Some(p) => p,
        None => {
            // 阶段内自动编译缺失二进制 (用户选项: 阶段内自动编译)
            tracing::info!(target: "sf_ocr", "未找到 subtitle-ocr, 尝试自动编译...");
            cargo_build_bin("subtitle-ocr-cli", "subtitle-ocr", &[], true).map_err(|e| {
                anyhow::anyhow!(
                    "{e}\n若编译失败, 请手动执行: cargo build --release -p subtitle-ocr-cli --bin subtitle-ocr"
                )
            })?
        }
    };

    let out_dir = sf_ocr_dir(&task_dir);
    std::fs::create_dir_all(&out_dir)
        .map_err(|e| anyhow::anyhow!("创建 {} 失败: {}", out_dir.display(), e))?;
    let out_file = out_dir.join("frames.json");

    let mut cmd = Command::new(&bin);
    cmd.arg("--dir").arg(&frame_dir);
    cmd.arg("--out").arg(&out_file);
    cmd.arg("--text-confidence-threshold");
    cmd.arg(cfg.text_confidence_threshold.to_string());
    if cfg.subtitle_only {
        cmd.arg("--subtitle-only");
    }

    tracing::info!(target: "sf_ocr", 
        "subtitle-ocr --dir {} --out {} --text-confidence-threshold {} {}",
        frame_dir.display(),
        out_file.display(),
        cfg.text_confidence_threshold,
        if cfg.subtitle_only {
            "--subtitle-only"
        } else {
            ""
        }
    );
    let status = cmd
        .status()
        .map_err(|e| anyhow::anyhow!("spawn subtitle-ocr 失败: {e}"))?;
    if !status.success() {
        return Err(anyhow::anyhow!(
            "subtitle-ocr failed with exit code {:?}",
            status.code()
        ));
    }

    let data: serde_json::Value = {
        let raw = std::fs::read_to_string(&out_file)
            .map_err(|e| anyhow::anyhow!("读取 {} 失败: {}", out_file.display(), e))?;
        serde_json::from_str(&raw)
            .map_err(|e| anyhow::anyhow!("解析 {} 失败: {}", out_file.display(), e))?
    };
    let frames = data
        .get("frames")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();
    if frames.is_empty() {
        return Err(anyhow::anyhow!("sf_ocr: no OCR results from keyframes"));
    }
    tracing::info!(target: "sf_ocr", 
        "{} frame results -> {}",
        frames.len(),
        out_file.display()
    );

    if cfg.cleanup_frames {
        let _ = std::fs::remove_dir_all(&frame_dir);
        tracing::info!(target: "sf_ocr", "Keyframes cleaned up");
    }

    set_stage_anyhow(
        &task_dir,
        "sf_ocr",
        StagePatch {
            status: Some(StageStatus::Success),
            completed_at: Some(now_iso()),
            progress: Some(100.0),
            last_message: Some(format!("OCR'd {} frame results", frames.len())),
            ..Default::default()
        },
    )?;
    tracing::info!(target: "sf_ocr", "sf_ocr: done");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::context::read_ctx_from_value;
    use serde_json::json;

    fn ctx_at(dir: &str, input: serde_json::Value) -> TaskCtx {
        let mut ctx = read_ctx_from_value(input).unwrap();
        ctx.task.task_dir = dir.to_string();
        ctx.pipeline = "dub".to_string();
        ctx
    }

    #[test]
    fn args_defaults() {
        let dir = "/x";
        let ctx = ctx_at(
            dir,
            json!({
                "task": {"id":"t","task_dir":dir,"url":"http://e","source":"remote",
                         "status":"running","created_at":"2024-01-01T00:00:00Z"},
                "input": {"stages": {"sf_ocr": {}}}
            }),
        );
        let cfg = read_args(&ctx);
        assert_eq!(cfg.text_confidence_threshold, 0.45);
        assert!(cfg.subtitle_only);
        assert!(!cfg.cleanup_frames);
    }

    #[test]
    fn args_camel_case() {
        let dir = "/x";
        let ctx = ctx_at(
            dir,
            json!({
                "task": {"id":"t","task_dir":dir,"url":"http://e","source":"remote",
                         "status":"running","created_at":"2024-01-01T00:00:00Z"},
                "input": {"stages": {"sf_ocr": {
                    "textConfidenceThreshold": 0.7, "subtitleOnly": false, "cleanupFrames": true
                }}}
            }),
        );
        let cfg = read_args(&ctx);
        assert_eq!(cfg.text_confidence_threshold, 0.7);
        assert!(!cfg.subtitle_only);
        assert!(cfg.cleanup_frames);
    }

    #[test]
    fn missing_frame_dir_errors() {
        let dir = std::env::temp_dir()
            .join(format!("ld_sfocr_{}", std::process::id()))
            .to_string_lossy()
            .to_string();
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let ctx = ctx_at(
            &dir,
            json!({
                "task": {"id":"t","task_dir":dir,"url":"http://e","source":"remote",
                         "status":"running","created_at":"2024-01-01T00:00:00Z"},
                "input": {}
            }),
        );
        crate::context::write_ctx(&dir, &ctx).unwrap();
        let res = stage_sf_ocr(&ctx);
        assert!(res.is_err());
        assert!(
            res.unwrap_err()
                .to_string()
                .contains("run sf_ocr_pre first"),
            "应提示先跑 sf_ocr_pre"
        );
        let _ = std::fs::remove_dir_all(&dir);
    }
}
